### PR TITLE
Add a basic pool manager for memory pools

### PR DESCRIPTION
**Pull request by kswiecicki**

### CODE REVIEW

File: source/common/ur_pool_manager.hpp [hunk 1]
```diff
@@ -29,8 +34,13 @@ struct pool_descriptor {
     ur_usm_type_t type;
     bool deviceReadOnly;
 
-    static bool equal(const pool_descriptor &lhs, const pool_descriptor &rhs);
-    static std::size_t hash(const pool_descriptor &desc);
+    struct hasher {
```
Comment (igchor) [inline source/common/ur_pool_manager.hpp:37]: nit: you can probably replace this with std::hash specialization

File: source/common/ur_pool_manager.hpp [hunk 2]
```diff
@@ -200,6 +203,74 @@ pool_descriptor::create(ur_usm_pool_handle_t poolHandle,
...
+                return {ret, pool_manager()};
+            }
+        }
+
+        return {UMA_RESULT_SUCCESS, std::move(manager)};
+    }
+
+    uma_result_t addPool(const D &desc,
+                         uma::pool_unique_handle_t &hPool) noexcept {
+        if (descToPoolMap.count(desc) != 0) {
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply source/common/ur_pool_manager.hpp]: Done.

File: source/common/ur_pool_manager.hpp [hunk 3]
```diff
@@ -122,22 +128,28 @@ isSharedAllocationReadOnlyOnDevice(const pool_descriptor &desc) {
...
 
     // We want to share a memory pool for sub-devices and sub-sub devices.
     // Sub-devices and sub-sub-devices might be represented by different ur_device_handle_t but
     // they share the same native_handle_t (which is used by UMF provider).
     // Ref: https://github.com/intel/llvm/commit/86511c5dc84b5781dcfd828caadcb5cac157eae1
     // TODO: is this L0 specific?
-    auto ret = urDeviceGetNativeHandle(lhs.hDevice, &lhsNative);
-    if (ret != UR_RESULT_SUCCESS) {
-        throw ret;
+    if (rhs.hDevice) {
```
Comment (igchor) [inline source/common/ur_pool_manager.hpp:141]: rhs -> lhs

File: test/usm/pool.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include <gtest/gtest.h>
+
+#include <umf/base.h>
+#include <umf/memory_provider.h>
+
+#include "umf_helpers.hpp"
+
+namespace usm_test {
+
+struct null_provider {
```
Comment (igchor) [inline test/usm/pool.hpp:23]: I think we already have null provider: https://github.com/oneapi-src/unified-runtime/blob/main/test/unified_malloc_framework/common/provider.h

Comment (kswiecicki) [reply test/usm/pool.hpp]: Done.

**Comment by pbalcer** [comment]
Ref: #517

File: test/usm/pool.hpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#include <gtest/gtest.h>
+
+#include <umf/base.h>
+#include <umf/memory_provider.h>
+
+#include "umf_helpers.hpp"
+
+namespace usm_test {
+
+struct null_pool {
```
Comment (igchor) [inline test/usm/pool.hpp:23]: I think we have null pool implementation already, as well.

Comment (igchor) [reply test/usm/pool.hpp:23]: https://github.com/oneapi-src/unified-runtime/blob/main/test/unified_malloc_framework/common/pool.h#L15

**Review by igchor — APPROVED** [review]

**Review by pbalcer — APPROVED** [review]

**Pull request merged**